CIP-148/README.md

```diff
@@ -0,0 +1,230 @@
+
+Other solutions involve enforcing that the index of the input in transaction inputs must correspond to the output at the same index of transaction outputs (ie `elemAt i txInfoInputs` corresponds to `elemAt i txInfoOutputs`. This really complicates transaction building, and often results in tons of fragmented UTxOs because the length of the `txInfoOutputs` list must be padded with filler UTxOs to match the indices of the required inputs. 
+
+The addition of `blake2b_224` in Plutus V3 enabled another design pattern to associate arbitrary data with transaction outputs. This design pattern can be used to prevent double satisfaction without sacrificing script composability. With the new `blake2b_224` primitive, we are able to hash things to the same format as `ScriptHash` and thus in a roundabout way we can validate against the contents of a reference scripts (which can be associated to outputs via the reference script field). Through this, we can associate arbitrary data to outputs without sacrificing script composability:
+```haskell
+spendValidator :: BuiltinData -> Integer -> ScriptContext -> () 
+spendValidator _ outIdx ctx = 
+ let ownInputRefHash = blake2b_224 (appendByteString (appendByteString "0100004881" (blake2b_224 (serialiseData (txOutRef ownInput)))) "0001")
+     Just ownOutputRefScriptHash = txOutReferenceScript ownOutput
+  in traceIfFalse "Output does not correspond to own input" (ownInputRefHash == ownOutputRefScriptHash) 
+ where
+    ownInput :: TxInInfo
+    ownInput = case findOwnInput ctx of
+        Nothing -> traceError "wrong script type / own input not found"
+        Just i  -> i
+    
+    ownOutput :: TxOut 
+    ownOutput = elemAt outIdx (txInfoOutputs (txInfo ctx))
+```
+_Note: you must prepend the Plutus version to "0100004881" for the above to work in production._
+
+When constructing a transaction to interact with the above validator we attach a reference script to each script output (for which we need to check correspondence). The reference script we attach is:
+```lisp
+(program 3.0.0 
+    (con bytestring #TX_OUT_REF_HASH)
+)
+```
+where `TX_OUT_REF_HASH` is the hash of the `TxOutRef` of the input for which this output corresponds to.  This trick offers a general solution to the double satisfaction problem but it doesn't solve the issues of dumping information that is only relevant during that one execution onto the chain forever. In-fact it only serves to exacerbate the issue, since we must store a reference script into the chain for each output that we want to associate arbitrary data to. 
+
+If this CIP is adopted, the above validator could be simplified to:
```

Comment:
This covers double satisfaction in the sense of multiple inputs being satisfied by one output. There is a dual problem - multiple outputs being satisfied by one input. To handle this, perhaps we should allow tags in transaction inputs, which can be checked by the guard scripts proposed in https://github.com/cardano-foundation/CIPs/pull/1063?